src/actors/objects/mod.rs

```diff
@@ -596,242 +233,235 @@ impl ObjectsActor {
-                                });
-                                Err(ObjectError::Caching(wrapped_error.error))
-                            }
-                        }
-                    })
-                    .filter(|response| match response {
-                        // Make sure we only return objects which provide the requested info
-                        Ok(object_meta) => {
-                            if object_meta.status == CacheStatus::Positive {
-                                // object_meta.features is meaningless when CacheStatus != Positive
-                                match purpose {
-                                    ObjectPurpose::Unwind => object_meta.features.has_unwind_info,
-                                    ObjectPurpose::Debug => {
-                                        object_meta.features.has_debug_info
-                                            || object_meta.features.has_symbols
-                                    }
-                                    ObjectPurpose::Source => object_meta.features.has_sources,
-                                }
-                            } else {
-                                true
-                            }
-                        }
-                        Err(_) => true,
-                    })
-                    .enumerate()
-                    .min_by_key(|(i, response)| {
-                        // The sources are ordered in priority, so for the same quality of
-                        // object file prefer a file from an earlier source using the index `i`
-                        // provided by enumerate.
-
-                        // Prefer files that contain an object over unparseable files
-                        let object_meta = match response {
-                            Ok(object_meta) => object_meta,
-                            Err(e) => {
-                                log::debug!("Failed to download: {:#?}", e);
-                                return (3, *i);
-                            }
-                        };
-
-                        // Prefer object files with debug/unwind info over object files without
-                        let score = match purpose {
-                            ObjectPurpose::Unwind if object_meta.features.has_unwind_info => 0,
-                            ObjectPurpose::Debug if object_meta.features.has_debug_info => 0,
-                            ObjectPurpose::Debug if object_meta.features.has_symbols => 1,
-                            ObjectPurpose::Source if object_meta.features.has_sources => 0,
-                            _ => 2,
-                        };
-                        (score, *i)
-                    })
-                    .map(|(_i, response)| response)
-                    .transpose()
-                    .map(|meta| FoundObject { meta, candidates: candidates.into() })
-            },
-        )
+            };
+            queries.push(query);
```

Comment:
you could keep the `file_ids.into_iter().map()`. Also, maybe in the future, `file_ids` could be a slice or an `&dyn Iterator` / `impl Iterator` instead of a `Vec`

Reply:
I think in the future this should become `fetch_file_meta()` which operates on just one file_id so that it can be chained on the future which creates the file_id and the join only happens at the end of this.  But I didn't want to change the logic for now and left this as a more pure async/await refactor.